### PR TITLE
In chat export, make sender name differentiable using `< >`

### PR DESCRIPTION
**Pull request by rameshsinghppc**
Fixes https://github.com/vector-im/element-web/issues/23837

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Signed-off-by: Ramesh Singh @rameshsinghppc

Type: [enhancement/defect/task] 

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * In chat export, make sender name differentiable using `< >` ([\#10748](https://github.com/matrix-org/matrix-react-sdk/pull/10748)). Fixes vector-im/element-web#23837. Contributed by @rameshsinghppc.

### CODE REVIEW

**Comment by richvdh** [comment]
This appears to be much the same as https://github.com/matrix-org/matrix-react-sdk/pull/10013, and as such has the same problems:

> This code is used throughout the whole product and not only in the context of chat exports.
> You will also need to write tests for this PR to be merged as per then contributing guidelines https://github.com/vector-im/element-web/blob/develop/CONTRIBUTING.md

Please read [CONTRIBUTING.md](https://github.com/vector-im/element-web/blob/develop/CONTRIBUTING.md) which gives instructions on how to make a good pull request. Looking at the format of other pull requests on the repository should help too.

**Review by richvdh — CHANGES_REQUESTED** [review]
^

**Comment by rameshsinghppc** [comment]
> Fixes [vector-im/element-web#23837](https://github.com/vector-im/element-web/issues/23837)

I have sent PR(https://github.com/matrix-org/matrix-js-sdk/pull/3345#issue-1698769126), please review it

**Comment by justjanne** [comment]
Both this and [the other new PR](https://github.com/matrix-org/matrix-js-sdk/pull/3345) are currently not in a state that can be reviewed or merged.

If you genuinely want to contribute this, I'd suggest adding another parameter to the functions in TextForEvent and using that to control whether the sender name will be wrapped in brackets. Take a look at the parameter that's used for controlling whether hidden events are included as an example.

**Comment by rameshsinghppc** [comment]
Screenshot from 2023-05-23 23-00-44

**Review by richvdh — CHANGES_REQUESTED** [review]
This is looking slightly more plausible but is causing a large number of type-checking and test failures. Please take the time to go through the check failures and address them.

File: src/TextForEvent.tsx [hunk 1]
```diff
@@ -942,8 +946,8 @@ export function hasText(ev: MatrixEvent, showHiddenEvents?: boolean): boolean {
  *     to avoid hitting the settings store
  */
 export function textForEvent(ev: MatrixEvent): string;
-export function textForEvent(ev: MatrixEvent, allowJSX: true, showHiddenEvents?: boolean): string | React.ReactNode;
-export function textForEvent(ev: MatrixEvent, allowJSX = false, showHiddenEvents?: boolean): string | React.ReactNode {
-    const handler = (ev.isState() ? stateHandlers : handlers)[ev.getType()];
-    return handler?.(ev, allowJSX, showHiddenEvents)?.() || "";
+export function textForEvent(ev: MatrixEvent, allowJSX: true, showHiddenEvents?: boolean,enclosedSender?:boolean): string | React.ReactNode;
+export function textForEvent(ev: MatrixEvent, allowJSX = false, showHiddenEvents?: boolean,enclosedSender?:boolean): string | React.ReactNode {
```
Comment (richvdh) [inline src/TextForEvent.tsx:950]: Please document the new parameter in the doc-comment above.

**Comment by richvdh** [comment]
@rameshsinghppc are you still keen to work on this? If you need pointers on how to fix up the CI failures then let us know.

**Comment by rameshsinghppc** [comment]
> @rameshsinghppc are you still keen to work on this? If you need pointers on how to fix up the CI failures then let us know.

@richvdh Yes, please help me to fix these failures.

**Comment by richvdh** [comment]
> @richvdh Yes, please help me to fix these failures.

Ok, what have you tried so far and which parts are confusing you?

**Comment by rameshsinghppc** [comment]
> > @richvdh Yes, please help me to fix these failures.
> 
> Ok, what have you tried so far and which parts are confusing you?

Screenshot from 2023-05-28 15-09-44

**Comment by richvdh** [comment]
Right; a lot of the problem here is that it is building against your fork of `matrix-js-sdk` (https://github.com/rameshsinghppc/matrix-js-sdk) which is outdated. (We have some magical branch matching.) You can see this happening at https://github.com/matrix-org/matrix-react-sdk/actions/runs/5159164960/jobs/9322532146#step:4:48.

Suggest, as a first step, you pull the latest `develop` into your fork of `matrix-js-sdk`. Ping me once you've done so and I'll re-run CI here.

**Comment by richvdh** [comment]
> Right; a lot of the problem here is that it is building against your fork of `matrix-js-sdk` (https://github.com/rameshsinghppc/matrix-js-sdk) which is outdated. (We have some magical branch matching.) You can see this happening at https://github.com/matrix-org/matrix-react-sdk/actions/runs/5159164960/jobs/9322532146#step:4:48.
> 
> Suggest, as a first step, you pull the latest `develop` into your fork of `matrix-js-sdk`. Ping me once you've done so and I'll re-run CI here.

To be more accurate here: it's actually building against [the branch named `rameshsingh` in your fork of `matrix-js-sdk`](https://github.com/rameshsinghppc/matrix-js-sdk/tree/rameshsingh). (Our magic branch-matching will use a js-sdk branch with the same name as your react-sdk branch).

So you need to update that branch (or just delete it).

**Comment by rameshsinghppc** [comment]
> > Right; a lot of the problem here is that it is building against your fork of `matrix-js-sdk` (https://github.com/rameshsinghppc/matrix-js-sdk) which is outdated. (We have some magical branch matching.) You can see this happening at https://github.com/matrix-org/matrix-react-sdk/actions/runs/5159164960/jobs/9322532146#step:4:48.
> > Suggest, as a first step, you pull the latest `develop` into your fork of `matrix-js-sdk`. Ping me once you've done so and I'll re-run CI here.
> 
> To be more accurate here: it's actually building against [the branch named `rameshsingh` in your fork of `matrix-js-sdk`](https://github.com/rameshsinghppc/matrix-js-sdk/tree/rameshsingh). (Our magic branch-matching will use a js-sdk branch with the same name as your react-sdk branch).
> 
> So you need to update that branch (or just delete it).

**Comment by rameshsinghppc** [comment]
> > Right; a lot of the problem here is that it is building against your fork of `matrix-js-sdk` (https://github.com/rameshsinghppc/matrix-js-sdk) which is outdated. (We have some magical branch matching.) You can see this happening at https://github.com/matrix-org/matrix-react-sdk/actions/runs/5159164960/jobs/9322532146#step:4:48.
> > Suggest, as a first step, you pull the latest `develop` into your fork of `matrix-js-sdk`. Ping me once you've done so and I'll re-run CI here.
> 
> To be more accurate here: it's actually building against [the branch named `rameshsingh` in your fork of `matrix-js-sdk`](https://github.com/rameshsinghppc/matrix-js-sdk/tree/rameshsingh). (Our magic branch-matching will use a js-sdk branch with the same name as your react-sdk branch).
> 
> So you need to update that branch (or just delete it).

@richvdh  Thank you for your help. I just deleted that repository.

**Comment by rameshsinghppc** [comment]
Screenshot from 2023-06-09 20-19-34

**Comment by richvdh** [comment]
you don't need to paste a screenshot - I can see the errors in CI.

Look at the line of code it is complaining about:

> else return textForEvent(mxEv, this.room.client,true) + mediaText;

What is the return type of `textForEvent`? (Hint: look at the [typescript definitions](https://github.com/matrix-org/matrix-react-sdk/pull/10748/files#diff-0e42a43897ea37a7954b73214502ba56f4d62b4adc509229c4fc483379102caeR964-L963) and consider which of the two prototypes your code is calling). Is it definitely a valid type to use in a string concatenation?

**Comment by richvdh** [comment]
@rameshsinghppc once you have made some changes you think will fix the CI please say so in a comment.

**Comment by rameshsinghppc** [comment]
> @rameshsinghppc once you have made some changes you think will fix the CI please say so in a comment.

@richvdh No sir, It's contributor's responsibility to resolve all the issues. This is my first open source contribution, so I don't know the steps for the open source contribution. I want to do open source contribution and learn through that.
One thing also I know that reviewer are very busy they don't have time for bad pull request but I am not getting any other ways through that I can learn open source contribution.
If you think my pull request is causing some problems for the maintainers then I am really sorry for that, you want then I can close this PR.

**Comment by rameshsinghppc** [comment]
@richvdh Few days back I made some changes in the code, please re- run the CI.

File: src/TextForEvent.tsx [hunk 2]
```diff
@@ -953,20 +958,29 @@ export function hasText(ev: MatrixEvent, client: MatrixClient, showHiddenEvents?
...
  *     to avoid hitting the settings store
+ * @param enclosedSender An optional parameter to check the sender name should be enclosed within
+ *      < > brackets or not
  */
 export function textForEvent(ev: MatrixEvent, client: MatrixClient): string;
+export function textForEvent(
+    ev: MatrixEvent,
+    client: MatrixClient,
+    enclosedSender: true,
+): string;
```
Comment (richvdh) [inline src/TextForEvent.tsx:969]: the problem here is that the new overload signature doesn't match the real definition of the function: the third parameter is always `allowJSX`.

https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads has more information on overload signatures, and explains it better than I can.

File: src/utils/exportUtils/PlainTextExport.ts [hunk 3]
```diff
@@ -103,8 +103,9 @@ export default class PlainTextExporter extends Exporter {
                 }
             } else mediaText = ` (${this.mediaOmitText})`;
         }
-        if (this.isReply(mxEv)) return senderDisplayName + ": " + this.textForReplyEvent(mxEv.getContent()) + mediaText;
-        else return textForEvent(mxEv, this.room.client) + mediaText;
+        if (this.isReply(mxEv))
+            return `<${senderDisplayName}>` + ": " + this.textForReplyEvent(mxEv.getContent()) + mediaText;
+        else return textForEvent(mxEv, this.room.client, true, true) + mediaText;
```
Comment (richvdh) [inline src/utils/exportUtils/PlainTextExport.ts:108]: This appears to set `allowJSX` and `showHiddenEvents` to `true` where neither was true before? Is that deliberate? If so, why?

**Review by richvdh — CHANGES_REQUESTED** [review]

**Comment by richvdh** [comment]
@rameshsinghppc there's no need to merge `develop` every day.

**Comment by richvdh** [comment]
Thanks for your contribution. I'm going to close this for now as the changes requested haven't been made. If you're interested in continuing to work on this please let us know and we can reopen the pull request. Thanks!

**Pull request closed without merging**